In [1]:
import pandas as pd
import pickle
from sklearn.mixture import GaussianMixture

In [2]:
def load(filename = "filename.pickle"): 
    try: 
        with open(filename, "rb") as file: 
            return pickle.load(file) 
    except FileNotFoundError: 
        print("File not found!") 

In [3]:
scaler2 = load("../Models/scaler.pickle")
scaler2

StandardScaler()

In [4]:
best_model = load("../Models/gmm4_6.pickle")

best_model

GaussianMixture(covariance_type='spherical', n_components=6)

In [5]:
hot_songs=pd.read_csv('../Data/hot_songs_final.csv')
hot_songs.head()

Unnamed: 0          title        artist  danceability  energy  key  \
0           0      Anti-Hero  Taylor Swift         0.637   0.643    4   
1           1     Lift Me Up       Rihanna         0.247   0.299    9   
2           2      Bad Habit    Steve Lacy         0.686   0.507    1   
3           3      As It Was  Harry Styles         0.520   0.731    6   
4           4  Lavender Haze  Taylor Swift         0.733   0.436   10   

   loudness  mode  speechiness  acousticness  ...  valence    tempo  \
0    -6.571     1       0.0519         0.130  ...   0.5330   97.008   
1    -6.083     1       0.0315         0.899  ...   0.1720  177.115   
2    -7.097     1       0.0357         0.626  ...   0.6900  168.955   
3    -5.338     0       0.0557         0.342  ...   0.6620  173.930   
4   -10.489     1       0.0800         0.258  ...   0.0976   96.985   

             type                      id  \
0  audio_features  0V3wPSX9ygBnCm8psDIegu   
1  audio_features  35ovElsgyAtQwYPYnZJECg   
2  audio_features  5CM4UuQ9Gnd6K2YyKGPMoK   
3  audio_features  4LRPiXqCikLlN15c3yImP7   
4  audio_features  5jQI2r1RdgtuT8S3iG8zFC   

                                    uri  \
0  spotify:track:0V3wPSX9ygBnCm8psDIegu   
1  spotify:track:35ovElsgyAtQwYPYnZJECg   
2  spotify:track:5CM4UuQ9Gnd6K2YyKGPMoK   
3  spotify:track:4LRPiXqCikLlN15c3yImP7   
4  spotify:track:5jQI2r1RdgtuT8S3iG8zFC   

                                          track_href  \
0  https://api.spotify.com/v1/tracks/0V3wPSX9ygBn...   
1  https://api.spotify.com/v1/tracks/35ovElsgyAtQ...   
2  https://api.spotify.com/v1/tracks/5CM4UuQ9Gnd6...   
3  https://api.spotify.com/v1/tracks/4LRPiXqCikLl...   
4  https://api.spotify.com/v1/tracks/5jQI2r1Rdgtu...   

                                        analysis_url duration_ms  \
0  https://api.spotify.com/v1/audio-analysis/0V3w...      200690   
1  https://api.spotify.com/v1/audio-analysis/35ov...      196520   
2  https://api.spotify.com/v1/audio-analysis/5CM4...      232115   
3  https://api.spotify.com/v1/audio-analysis/4LRP...      167303   
4  https://api.spotify.com/v1/audio-analysis/5jQI...      202396   

  time_signature  set  
0              4    H  
1              4    H  
2              4    H  
3              4    H  
4              4    H  

[5 rows x 22 columns]

In [6]:
not_hot_songs=pd.read_csv('../Data/not_hot_songs_final.csv')
not_hot_songs.head()

Unnamed: 0                           title              artist  \
0           0            Like a Rolling Stone           Bob Dylan   
1           1  (I Can't Get No) Satisfaction'  The Rolling Stones   
2           2                         Imagine         John Lennon   
3           3                 What's Going On         Marvin Gaye   
4           4                         Respect     Aretha Franklin   

   danceability  energy  key  loudness  mode  speechiness  acousticness  ...  \
0         0.482   0.721    0    -6.839     1       0.0321        0.7310  ...   
1         0.723   0.863    2    -7.890     1       0.0338        0.0383  ...   
2         0.547   0.257    0   -12.358     1       0.0252        0.9070  ...   
3         0.283   0.716    1    -9.632     0       0.0986        0.4470  ...   
4         0.805   0.558    0    -5.226     1       0.0410        0.1640  ...   

   valence    tempo            type                      id  \
0    0.557   95.263  audio_features  3AhXZa8sUQht0UEdBJgpGc   
1    0.931  136.302  audio_features  2PzU4IB8Dr6mxV3lHuaG34   
2    0.169   75.752  audio_features  7pKfPomDEeI4TPT6EOYjn9   
3    0.828  201.960  audio_features  3Um9toULmYFGCpvaIPFw7l   
4    0.965  114.950  audio_features  7s25THrKz86DM225dOYwnr   

                                    uri  \
0  spotify:track:3AhXZa8sUQht0UEdBJgpGc   
1  spotify:track:2PzU4IB8Dr6mxV3lHuaG34   
2  spotify:track:7pKfPomDEeI4TPT6EOYjn9   
3  spotify:track:3Um9toULmYFGCpvaIPFw7l   
4  spotify:track:7s25THrKz86DM225dOYwnr   

                                          track_href  \
0  https://api.spotify.com/v1/tracks/3AhXZa8sUQht...   
1  https://api.spotify.com/v1/tracks/2PzU4IB8Dr6m...   
2  https://api.spotify.com/v1/tracks/7pKfPomDEeI4...   
3  https://api.spotify.com/v1/tracks/3Um9toULmYFG...   
4  https://api.spotify.com/v1/tracks/7s25THrKz86D...   

                                        analysis_url duration_ms  \
0  https://api.spotify.com/v1/audio-analysis/3AhX...      369600   
1  https://api.spotify.com/v1/audio-analysis/2PzU...      222813   
2  https://api.spotify.com/v1/audio-analysis/7pKf...      187867   
3  https://api.spotify.com/v1/audio-analysis/3Um9...      233000   
4  https://api.spotify.com/v1/audio-analysis/7s25...      147600   

  time_signature  set  
0              4    N  
1              4    N  
2              4    N  
3              4    N  
4              4    N  

[5 rows x 22 columns]

In [7]:
from config import *

In [8]:
import spotipy
import json
from spotipy.oauth2 import SpotifyClientCredentials

#Initialize SpotiPy with user credentias
sp = spotipy.Spotify(auth_manager=SpotifyClientCredentials(client_id= Client_ID,
                                                           client_secret= Client_Secret))


In [9]:
whole_set = pd.concat([not_hot_songs, hot_songs], axis=0) 


In [10]:
whole_set_copy= pd.concat([not_hot_songs, hot_songs], axis=0) 
whole_set_copy=whole_set_copy.drop(columns = ['Unnamed: 0', 'uri', 'analysis_url'], axis=1).reset_index(drop=True)
whole_set_copy #this df has a set, title, and link to a song

title              artist  danceability  energy  \
0              Like a Rolling Stone           Bob Dylan         0.482   0.721   
1    (I Can't Get No) Satisfaction'  The Rolling Stones         0.723   0.863   
2                           Imagine         John Lennon         0.547   0.257   
3                   What's Going On         Marvin Gaye         0.283   0.716   
4                           Respect     Aretha Franklin         0.805   0.558   
..                              ...                 ...           ...     ...   
555               Betty (Get Money)          Yung Gravy         0.723   0.772   
556                    300 Blackout         Kodak Black         0.669   0.741   
557                     Dear Reader        Taylor Swift         0.562   0.388   
558                      Country On          Luke Bryan         0.520   0.751   
559                       Evergreen         Omar Apollo         0.697   0.336   

     key  loudness  mode  speechiness  acousticness  instrumentalness  \
0      0    -6.839     1       0.0321       0.73100          0.000000   
1      2    -7.890     1       0.0338       0.03830          0.031700   
2      0   -12.358     1       0.0252       0.90700          0.183000   
3      1    -9.632     0       0.0986       0.44700          0.000000   
4      0    -5.226     1       0.0410       0.16400          0.000022   
..   ...       ...   ...          ...           ...               ...   
555    1    -6.784     1       0.0792       0.00524          0.000012   
556    0    -6.545     0       0.3160       0.01980          0.000000   
557    0   -12.088     1       0.0705       0.48100          0.001450   
558    5    -5.064     1       0.0551       0.17100          0.000007   
559    9   -10.395     1       0.0424       0.59800          0.005280   

     liveness  valence    tempo            type                      id  \
0      0.1890    0.557   95.263  audio_features  3AhXZa8sUQht0UEdBJgpGc   
1      0.1280    0.931  136.302  audio_features  2PzU4IB8Dr6mxV3lHuaG34   
2      0.0935    0.169   75.752  audio_features  7pKfPomDEeI4TPT6EOYjn9   
3      0.3990    0.828  201.960  audio_features  3Um9toULmYFGCpvaIPFw7l   
4      0.0546    0.965  114.950  audio_features  7s25THrKz86DM225dOYwnr   
..        ...      ...      ...             ...                     ...   
555    0.3670    0.580  102.004  audio_features  4cacyP5c3PMlfnyjpg13xW   
556    0.2230    0.396  167.903  audio_features  3y3NwQPQAjKCunRz2eKggA   
557    0.1170    0.159  107.747  audio_features  3QF5RsWzK1lCvf2o2cY65P   
558    0.0624    0.519  156.044  audio_features  1tRxwf8Q0AcshfHuaD86Yt   
559    0.1070    0.310   82.029  audio_features  2TktkzfozZifbQhXjT6I33   

                                            track_href  duration_ms  \
0    https://api.spotify.com/v1/tracks/3AhXZa8sUQht...       369600   
1    https://api.spotify.com/v1/tracks/2PzU4IB8Dr6m...       222813   
2    https://api.spotify.com/v1/tracks/7pKfPomDEeI4...       187867   
3    https://api.spotify.com/v1/tracks/3Um9toULmYFG...       233000   
4    https://api.spotify.com/v1/tracks/7s25THrKz86D...       147600   
..                                                 ...          ...   
555  https://api.spotify.com/v1/tracks/4cacyP5c3PMl...       146471   
556  https://api.spotify.com/v1/tracks/3y3NwQPQAjKC...       166494   
557  https://api.spotify.com/v1/tracks/3QF5RsWzK1lC...       225194   
558  https://api.spotify.com/v1/tracks/1tRxwf8Q0Acs...       236456   
559  https://api.spotify.com/v1/tracks/2TktkzfozZif...       216680   

     time_signature set  
0                 4   N  
1                 4   N  
2                 4   N  
3                 4   N  
4                 4   N  
..              ...  ..  
555               4   H  
556               4   H  
557               4   H  
558               4   H  
559               3   H  

[560 rows x 19 columns]

In [11]:
whole_set=whole_set.drop(columns = ['Unnamed: 0', 'title', 'artist','type', 'key', 'mode', 'id', 'uri', 'track_href',
       'analysis_url', 'duration_ms', 'time_signature', 'set'], axis=1).reset_index(drop=True)
whole_set #this data frame does not have a set

danceability  energy  loudness  speechiness  acousticness  \
0           0.482   0.721    -6.839       0.0321       0.73100   
1           0.723   0.863    -7.890       0.0338       0.03830   
2           0.547   0.257   -12.358       0.0252       0.90700   
3           0.283   0.716    -9.632       0.0986       0.44700   
4           0.805   0.558    -5.226       0.0410       0.16400   
..            ...     ...       ...          ...           ...   
555         0.723   0.772    -6.784       0.0792       0.00524   
556         0.669   0.741    -6.545       0.3160       0.01980   
557         0.562   0.388   -12.088       0.0705       0.48100   
558         0.520   0.751    -5.064       0.0551       0.17100   
559         0.697   0.336   -10.395       0.0424       0.59800   

     instrumentalness  liveness  valence    tempo  
0            0.000000    0.1890    0.557   95.263  
1            0.031700    0.1280    0.931  136.302  
2            0.183000    0.0935    0.169   75.752  
3            0.000000    0.3990    0.828  201.960  
4            0.000022    0.0546    0.965  114.950  
..                ...       ...      ...      ...  
555          0.000012    0.3670    0.580  102.004  
556          0.000000    0.2230    0.396  167.903  
557          0.001450    0.1170    0.159  107.747  
558          0.000007    0.0624    0.519  156.044  
559          0.005280    0.1070    0.310   82.029  

[560 rows x 9 columns]

In [12]:
cols = ['danceability', 'energy', 'loudness', 'speechiness', 'acousticness',
       'instrumentalness', 'liveness', 'valence', 'tempo'] 
#these a columns that are needed to scaler and for gaussian mixture

In [13]:
set_features_scaled_np = scaler2.transform(whole_set)
set_features_scaled_df = pd.DataFrame(set_features_scaled_np, columns=whole_set.columns)

In [14]:
ygmm_hat4 = best_model.predict(set_features_scaled_df[cols])


In [15]:
set_features_scaled_df["cluster_spheric"] = ygmm_hat4
set_features_scaled_df.head()

danceability    energy  loudness  speechiness  acousticness  \
0     -0.556691  0.598239  0.672181    -0.491641      1.320872   
1      0.976657  1.281999  0.361757    -0.469691     -1.042073   
2     -0.143132 -1.636020 -0.957914    -0.580732      1.921245   
3     -1.822817  0.574163 -0.152762     0.366996      0.352089   
4      1.498378 -0.186641  1.148597    -0.376726     -0.613284   

   instrumentalness  liveness   valence     tempo  cluster_spheric  
0         -0.292743 -0.038574 -0.230137 -0.942883                4  
1         -0.077030 -0.387415  1.312121  0.477670                1  
2          0.952540 -0.584710 -1.830126 -1.618250                5  
3         -0.292743  1.162355  0.887381  2.750403                3  
4         -0.292594 -0.807168  1.452326 -0.261423                1

In [16]:
all_songs_clustered=pd.concat([set_features_scaled_df, whole_set_copy['set'],whole_set_copy['title'],whole_set_copy['artist']], axis=1).reset_index(drop=True)
#all_songs_clustered
all_songs_clustered

danceability    energy  loudness  speechiness  acousticness  \
0       -0.556691  0.598239  0.672181    -0.491641      1.320872   
1        0.976657  1.281999  0.361757    -0.469691     -1.042073   
2       -0.143132 -1.636020 -0.957914    -0.580732      1.921245   
3       -1.822817  0.574163 -0.152762     0.366996      0.352089   
4        1.498378 -0.186641  1.148597    -0.376726     -0.613284   
..            ...       ...       ...          ...           ...   
555      0.976657  0.843815  0.688425     0.116506     -1.154847   
556      0.633086  0.694543  0.759017     3.174026     -1.105180   
557     -0.047696 -1.005227 -0.878167     0.004173      0.468070   
558     -0.314918  0.742695  1.196446    -0.194669     -0.589405   
559      0.811234 -1.255619 -0.378122    -0.358649      0.867181   

     instrumentalness  liveness   valence     tempo  cluster_spheric set  \
0           -0.292743 -0.038574 -0.230137 -0.942883                4   N   
1           -0.077030 -0.387415  1.312121  0.477670                1   N   
2            0.952540 -0.584710 -1.830126 -1.618250                5   N   
3           -0.292743  1.162355  0.887381  2.750403                3   N   
4           -0.292594 -0.807168  1.452326 -0.261423                1   N   
..                ...       ...       ...       ...              ...  ..   
555         -0.292665  0.979356 -0.135292 -0.709545                1   H   
556         -0.292743  0.155862 -0.894050  1.571530                0   H   
557         -0.282876 -0.450321 -1.871363 -0.510753                5   H   
558         -0.292693 -0.762562 -0.386837  1.161034                3   H   
559         -0.256814 -0.507508 -1.248687 -1.400974                4   H   

                              title              artist  
0              Like a Rolling Stone           Bob Dylan  
1    (I Can't Get No) Satisfaction'  The Rolling Stones  
2                           Imagine         John Lennon  
3                   What's Going On         Marvin Gaye  
4                           Respect     Aretha Franklin  
..                              ...                 ...  
555               Betty (Get Money)          Yung Gravy  
556                    300 Blackout         Kodak Black  
557                     Dear Reader        Taylor Swift  
558                      Country On          Luke Bryan  
559                       Evergreen         Omar Apollo  

[560 rows x 13 columns]

The following function gets a dataframe with our 560 songs as an input. It asks the customer to provide title and artist,after it gets audio features of this song, finds to which cluster it belongs. We also check if song belongs to Hot Song or Not Hot Song.
If it's a hot song we provide a recommendation from the same cluster and the list of hot songs
In other case, any song can be recommended but from the whole set and the same cluster


In [17]:
def search_url(title):
    query = title
    search = sp.search(q=query, type="track")
    uri = search["tracks"]["items"][0]['external_urls']
    return uri

In [33]:
def find_song(df):
    sample=pd.DataFrame()  
    df=df.copy()
       #Get name of song and artist from the client!
    fav_song = input('What is your favourite song?\n')
    print('______________________________________________________')
    singer=input('What is the artist name?\n')
    print('______________________________________________________')
    print('______________________________________________________')
    print('______________________________________________________')
    
    string = "track:" + fav_song + " artist:" + singer 
    result = sp.search(q=string, limit=5)
    #print(result)
    
    if len(result['tracks']['items']) == 0:
        print("Sorry, your song was not found!")
        sample = "No song"
        url = None
        return sample, url
    else: 
              
        song_id = result['tracks']['items'][0]['id'] #here we get the user song id
    #getting features
        features = sp.audio_features((song_id))[0]      
        dict_features = { key : [features[key]] for key in list(features.keys()) }   #making a dictionary
        features_df=pd.DataFrame(dict_features)                                #making a data frame          
        
        features_df=features_df.drop(columns = ['type', 'key', 'mode', 'id', 'uri', 'track_href',
           'analysis_url', 'duration_ms', 'time_signature'], axis=1)   #removing columns that not needed for clustering
        
    
    #scaling features, loading model and getting the user_cluster
        features_scaled_np = scaler2.transform(features_df)
        features_scaled_df = pd.DataFrame(features_scaled_np, columns=features_df.columns)
        best_model = load("../Models/gmm4_6.pickle")
        
        user_cluster = best_model.predict(features_scaled_df)[0]
     
        if (len(result['tracks']['items'])> 1):
            if song_id in list(hot_songs['id']):
                    
                      
                sample = df[(df['set']=="H") & (df['cluster_spheric']==user_cluster)].sample()
                #print("First if ",sample)
                url = search_url(sample['title'])
                print('We recommend you this song:')
            
            elif song_id in list(not_hot_songs['id']):
            
                sample = df[(df['set']=="N") & (df['cluster_spheric']==user_cluster)].sample()
                #print("elif ", sample)
                url = search_url(sample['title'])
                print('We recommend you this song:')
        
            else: 
            
                sample = df[(df['set']=="N") & (df['cluster_spheric']==user_cluster)].sample()
                #print("else ",sample)
                url = search_url(sample['title'])
                print('We recommend you this song:')
        else:
            sample = df[(df['set']=="N") & (df['cluster_spheric']==user_cluster)].sample()
                #print("else ",sample)
            url = search_url(sample['title'])
            print('We recommend you this song:')
            
           
        
        return sample, url 

       

In [44]:
sample, url = find_song(all_songs_clustered)

if (type(sample) == pd.DataFrame):
    print('Title of the song:', sample['title'].to_string(index=False))
    print('Artist name:', sample['artist'].to_string(index=False))
    print('Click on this link: ') 
    print(url['spotify'])

What is your favourite song?
Redemption Song
______________________________________________________
What is the artist name?
Bob Marley
______________________________________________________
______________________________________________________
______________________________________________________
We recommend you this song:
Title of the song: The Tracks of My Tears
Artist name: Smokey Robinson and the Miracles
Click on this link: 
https://open.spotify.com/track/6QyQmdvQ1ywNccYa0pwLNQ
